doc/design/networking/Investigation-on-NSX-integration.md

```diff
@@ -0,0 +1,109 @@
+
+## NSX Manager API:
+![Code flow to use NSX Manager API] (pics/with-nsx.png)
+
+### API endpoint
+https://{nsxmanager's hostname or IP}/api 
+
+### API calls
+(details in http://pubs.vmware.com/nsx-63/topic/com.vmware.ICbase/PDF/nsx_63_api.pdf , note that many NSX API methods reference vCenter object IDs in URI parameters, query parameters, request bodies, and response bodies, also it takes XML as request body (sad) )
+* docker network create
+  * create IP Sets
+     * POST /2.0/services/ipset/{scopeMoref}
+  * or create Security Tag
+     * POST /2.0/services/securitytags/tag
+  * create security group withe IP Sets or Security Tag as members
+     * POST /2.0/services/securitygroup/bulk/{scopeId} (For the scopeId use globalroot-0 for non-universal security groups and universalroot-0 for universal security groups.)
+  * create security policy and apply it to security group
+     * POST /2.0/services/policy/securitypolicy (specify securityGroupBinding in the xml request body)
+
+* docker network delete 
+  * check if any VMs attached to this security group
+     * GET /2.0/services/securitygroup/{objectId}/translation/virtualmachine 
+  * delete the network (security group, policy group, IP Sets)
+     * DELETE /2.0/services/securitytags/tag/{tagId}
+     * DELETE /2.0/services/securitygroup/{objectId} 
+     * DELETE /2.0/services/ipset/{ipsetId}
+
+* docker network list
+  * list all the security groups
+    * GET /2.0/services/securitygroup/internal/scope/{scopeId}
+  * filter security group names by `<vch name>-SecurityGroup` prefix
+  * or if we keep KV pairs of each created network, then we just need to use that to list all the networks
+
+* docker run <image> --net=<network-name>
+  * create a containerVM
+  * if using security tag as the membership condition, need to assign security tag to the containerVM
+    * POST /2.0/services/securitytags/vm/{vmId}
+
+## State Storage
+  * we can map a docker network name to a security group name as `<vch name>-SecurityGroup-<network name>`
+  * we can store KV pairs for each created network and then everything else will be in the NSX control plane
+
+## Unclear / need discussion:
+### Security Tag v.s IP Sets
+  * Security Tag we only need to add the tag to a VM then it is added to a security group, which makes adding VMs to multiple security groups convinient. We do not need to update security groups everytime we create a new containerVM in a network.
+  * With IP Sets, we need to add the VM's IP to an IP Set. The downside is that 1. it requires the VM's IP to be static. 2. if we have more than one vch in the environment, containerVMs' IPs overlap, we cannot allow one VM in multiple security groups. 3. it may limit the number of VMs added to an IP Set (how many IPs can be added one IP set?). 4. everytime adding a new containerVM in a network, we need to add its IP to the IP set.
```

Comment:
Please reflow the numbered points so they render as a list.

Consideration of security tag:
1. Seems to apply to a VM instead of an interface on a VM - does this allow for different policies per vNIC?

Other possible considerations of IPs:
5. currently we only assign an IP to a VM when it is powered on so VMs wouldn't have a security group membership when powered off (meaning possible issues with out-of-band power on)
6. no way to prevent a container changing it's IP address